File: project_euler/src/problem_051.erl
%% @author std-string

%% By replacing the 1-st digit of the 2-digit number *3, it turns out that six of the nine possible values: 13, 23, 43, 53, 73, and 83, are all prime.
%% By replacing the 3-rd and 4-th digits of 56**3 with the same digit, this 5-digit number is the first example having seven primes among the ten generated numbers,
%% yielding the family: 56003, 56113, 56333, 56443, 56663, 56773, and 56993. Consequently 56003, being the first member of this family,
%% is the smallest prime with this property.
%% Find the smallest prime which, by replacing part of the number (not necessarily adjacent digits) with the same digit, is part of an eight prime value family.

-module(problem_051).
-export([get_check_data/0, prepare_data/2, solve/1]).

-behaviour(numerical_task_behaviour).

-type digit_type() :: 0..9.

-define(MAX_NUMBER, 100000000-1).
-define(RANGE_START, 10).
-define(DIGITS_COUNT, 2).

%% ====================================================================
%% API functions
%% ====================================================================

get_check_data() ->
    [{7, 56003}, {8, 121313}].

prepare_data(_ModuleSourceDir, Input) -> Input.

solve(FamilySize) ->
    Primes = eratos_sieve:get_sieve(?MAX_NUMBER),
    case process_numbers(?RANGE_START, ?DIGITS_COUNT, FamilySize, Primes) of
        {true, Prime} -> Prime;
        false -> error(solution_not_found)
    end.

%% ====================================================================
%% Internal functions
%% ====================================================================

-spec create_digits_info(Number :: pos_integer(), DigitsCount :: pos_integer()) -> array:array([non_neg_integer()]).
create_digits_info(Number, DigitsCount) ->
    Digits = numbers:get_digits(Number),
    InfoInit = array:new([{size, 10}, {fixed, true}, {default, []}]),
    AccInit = {InfoInit, DigitsCount - 1},
    {InfoFinal, _} = lists:foldl(fun(Digit, {Info, Pos}) -> {array:set(Digit, [Pos] ++ array:get(Digit, Info), Info), Pos - 1} end, AccInit, Digits),
    InfoFinal.

%%-spec select_possible_variants(DigitsInfo :: array:array([non_neg_integer()]), FamilySize :: pos_integer()) ->
%%    [{Digit :: digit_type(), PosList :: [non_neg_integer()]}].
%%select_possible_variants(DigitsInfo, FamilySize) -> select_possible_variants(DigitsInfo, FamilySize, 0, []).

%%-spec select_possible_variants(DigitsInfo :: array:array([non_neg_integer()]),
%%                               FamilySize :: pos_integer(),
%%                               Digit :: digit_type(),
%%                               Storage :: [{Digit :: digit_type(), PosList :: [non_neg_integer()]}]) ->
%%    [{Digit :: digit_type(), PosList :: [non_neg_integer()]}].
%%select_possible_variants(_DigitsInfo, _FamilySize, Digit, Storage) when Digit > 9 -> Storage;
%%select_possible_variants(DigitsInfo, FamilySize, Digit, Storage) ->
%%    PosList = array:get(Digit, DigitsInfo),
%%    case check_possible_variant(PosList, Digit, FamilySize) of
%%        false -> select_possible_variants(DigitsInfo, FamilySize, Digit + 1, Storage);
%%        true -> select_possible_variants(DigitsInfo, FamilySize, Digit + 1, [{Digit, PosList}] ++ Storage)
%%    end.

-spec select_possible_variants(DigitsInfo :: array:array([non_neg_integer()]), FamilySize :: pos_integer()) -> [digit_type()].
select_possible_variants(DigitsInfo, FamilySize) -> select_possible_variants(DigitsInfo, FamilySize, 0, []).

-spec select_possible_variants(DigitsInfo :: array:array([non_neg_integer()]),
                               FamilySize :: pos_integer(),
                               Digit :: digit_type(),
                               Storage :: [digit_type()]) -> [digit_type()].
select_possible_variants(_DigitsInfo, _FamilySize, Digit, Storage) when Digit > 9 -> Storage;
select_possible_variants(DigitsInfo, FamilySize, Digit, Storage) ->
    PosList = array:get(Digit, DigitsInfo),
    case check_possible_variant(PosList, Digit, FamilySize) of
        false -> select_possible_variants(DigitsInfo, FamilySize, Digit + 1, Storage);
        true -> select_possible_variants(DigitsInfo, FamilySize, Digit + 1, [Digit] ++ Storage)
    end.

-spec check_possible_variant(PosList :: [non_neg_integer()], Digit :: digit_type(), FamilySize :: pos_integer()) -> boolean().
check_possible_variant([], _Digit, _FamilySize) -> false;
%% we don't process 1-digit family
check_possible_variant([_Pos], _Digit, _FamilySize) -> false;
%% last (younger) digit must be 1, 3, 7, 9
check_possible_variant([0 | _Rest], 1, FamilySize) -> FamilySize =< 4;
check_possible_variant([0 | _Rest], 3, FamilySize) -> FamilySize =< 3;
check_possible_variant([0 | _Rest], 7, FamilySize) -> FamilySize =< 2;
check_possible_variant([0 | _Rest], 9, FamilySize) -> FamilySize =< 1;
check_possible_variant(_PosList, Digit, FamilySize) -> FamilySize =< 10 - Digit.

-spec calc_number_part(Digit :: digit_type(), PosList :: [non_neg_integer()]) -> non_neg_integer().
calc_number_part(0, _PosList) -> 0;
calc_number_part(_Digit, []) -> 0;
calc_number_part(Digit, PosList) ->
    %% TODO (std_string) : probably optimize this place
    Digit * lists:sum(lists:map(fun(Pos) -> numbers:power(10, Pos) end, PosList)).

-spec create_number(DigitsInfo :: array:array([non_neg_integer()])) -> non_neg_integer().
create_number(DigitsInfo) ->
    array:foldl(fun(Digit, PosList, Number) -> calc_number_part(Digit, PosList) + Number end, 0, DigitsInfo).

-spec check_family(DigitsInfo :: array:array([non_neg_integer()]),
                   Digit :: digit_type(),
                   FamilySize :: pos_integer(),
                   Primes :: eratos_sieve:sieve()) -> boolean().
check_family(DigitsInfo, Digit, FamilySize, Primes) ->
    PosList = array:get(Digit, DigitsInfo),
    InvDigitsInfo = array:set(Digit, [], DigitsInfo),
    InvNumberPart = create_number(InvDigitsInfo),
    case PosList of
        [0 | _Rest] ->
            %% TODO (std_string) : move [1, 3, 7, 9] into definitions
            Digits = lists:dropwhile(fun(Number) -> Number =< Digit end, [1, 3, 7, 9]),
            check_family_impl(InvNumberPart, Digits, PosList, FamilySize - 1, Primes);
        _ ->
            Digits = lists:seq(Digit + 1, 9),
            check_family_impl(InvNumberPart, Digits, PosList, FamilySize - 1, Primes)
    end.

-spec check_family_impl(InvNumberPart :: non_neg_integer(),
                        Digits :: [digit_type()],
                        PosList :: [non_neg_integer()],
                        FamilySizeRest :: integer(),
                        Primes :: eratos_sieve:sieve()) -> boolean().
check_family_impl(_InvNumberPart, Digits, _PosList, FamilySizeRest, _Primes) when length(Digits) < FamilySizeRest -> false;
%%check_family_impl(_InvNumberPart, [], _PosList, FamilySizeRest, _Primes) when FamilySizeRest <= 0 -> true;
check_family_impl(_InvNumberPart, [], _PosList, 0, _Primes) -> true;
check_family_impl(InvNumberPart, [Digit | DigitsRest], PosList, FamilySizeRest, Primes) ->
    Number = InvNumberPart + calc_number_part(Digit, PosList),
    case eratos_sieve:is_prime(Number, Primes) of
        true -> check_family_impl(InvNumberPart, DigitsRest, PosList, FamilySizeRest - 1, Primes);
        false -> check_family_impl(InvNumberPart, DigitsRest, PosList, FamilySizeRest, Primes)
    end.

-spec check_possible_families(DigitsInfo :: array:array([non_neg_integer()]),
                              Variants :: [digit_type()],
                              FamilySize :: pos_integer(),
                              Primes :: eratos_sieve:sieve()) -> boolean().
check_possible_families(_DigitsInfo, [], _FamilySize, _Primes) -> false;
check_possible_families(DigitsInfo, [Variant | VariantsRest], FamilySize, Primes) ->
    case check_family(DigitsInfo, Variant, FamilySize, Primes) of
        false -> check_possible_families(DigitsInfo, VariantsRest, FamilySize, Primes);
        true -> true
    end.

-spec process_range(Current :: pos_integer(),
                    Max :: pos_integer(),
                    DigitsCount :: pos_integer(),
                    FamilySize :: pos_integer(),
                    Primes :: eratos_sieve:sieve()) -> {'true', Number :: pos_integer()} | 'false'.
process_range(Current, Max, _DigitsCount, _FamilySize, _Primes) when Current > Max -> false;
process_range(Current, Max, DigitsCount, FamilySize, Primes) ->
    case eratos_sieve:is_prime(Current, Primes) of
        true ->
            DigitsInfo = create_digits_info(Current, DigitsCount),
            PossibleVariants = select_possible_variants(DigitsInfo, FamilySize),
            case check_possible_families(DigitsInfo, PossibleVariants, FamilySize, Primes) of
                false -> process_range(Current + 2, Max, DigitsCount, FamilySize, Primes);
                true -> {true, Current}
            end;
        false -> process_range(Current + 2, Max, DigitsCount, FamilySize, Primes)
    end.

-spec process_numbers(RangeStart :: pos_integer(),
                      DigitsCount :: pos_integer(),
                      FamilySize :: pos_integer(),
                      Primes :: eratos_sieve:sieve()) -> {'true', Number :: pos_integer()} | 'false'.
process_numbers(RangeStart, _DigitsCount, _FamilySize, _Primes) when RangeStart > ?MAX_NUMBER -> false;
process_numbers(RangeStart, DigitsCount, FamilySize, Primes) ->
    RangeFinish = 10 * RangeStart - 3,
    case process_range(RangeStart + 1, RangeFinish, DigitsCount, FamilySize, Primes) of
        false -> process_numbers(10 * RangeStart, DigitsCount + 1, FamilySize, Primes);
        {true, Prime} -> {true, Prime}
    end.

File: projecteuler.fsharp/ProjectEulerTasks/Problem051.fs
namespace ProjectEulerTasks

open NUnit.Framework
open ProjectEulerTasks.Utils
open CommonLib

// By replacing the 1-st digit of the 2-digit number *3, it turns out that six of the nine possible values: 13, 23, 43, 53, 73, and 83, are all prime.
// By replacing the 3-rd and 4-th digits of 56**3 with the same digit, this 5-digit number is the first example having seven primes among the ten generated numbers,
// yielding the family: 56003, 56113, 56333, 56443, 56663, 56773, and 56993. Consequently 56003, being the first member of this family, is the smallest prime with this property.
// Find the smallest prime which, by replacing part of the number (not necessarily adjacent digits) with the same digit, is part of an eight prime value family.

module Problem051Impl =

    type DigitInfo = {Digit: int; Positions: int list}

    type DigitsInfo = {Info: DigitInfo[]; ElderPos: int}

open Problem051Impl

[<TestFixture>]
type Problem051() =

    // TODO (std_string) : think about selection of this value
    [<Literal>]
    let MaxNumber = 9999999

    // TODO (std_string) : think about selection of this value
    [<Literal>]
    let MaxDigitsCount = 7

    let digitsWeight = seq {0 .. MaxDigitsCount - 1} |> Seq.map (fun digitPos -> pown 10 digitPos) |> Seq.toArray

    let createDigitsInfo (number: int) =
        let digitsCount = 10
        let digitsInfo = Array.create digitsCount []
        let digits = number |> NumbersDigits.GetDigits
        digits |> Seq.rev |> Seq.iteri (fun pos digit -> digitsInfo.[digit] <- pos :: digitsInfo.[digit])
        {DigitsInfo.Info = seq {0 .. 9} |> Seq.map (fun digit -> {DigitInfo.Digit = digit; DigitInfo.Positions = digitsInfo.[digit]}) |> Seq.toArray; DigitsInfo.ElderPos = digits.Length - 1}

    let createDigitPart (digitInfo: DigitInfo) =
        digitInfo.Positions |> Seq.map (fun pos -> digitsWeight.[pos]) |> Seq.sum |> (*) digitInfo.Digit

    let createNumberBasis (excludedDigit: int) (digitsInfo: DigitsInfo) =
        seq {0 .. 9} |> Seq.filter (fun digit -> digit <> excludedDigit) |> Seq.map (fun digit -> digitsInfo.Info.[digit] |> createDigitPart) |> Seq.sum

    let createNumber (numberBasis: int) (variantDigitInfo: DigitInfo) =
        numberBasis + (variantDigitInfo |> createDigitPart)

    let checkPossibleVariant (familySize: int) (elderPos: int) (digitInfo: DigitInfo) =
        match digitInfo.Digit, digitInfo.Positions with
        | _, [] -> false
        // last (younger) digit must be 1, 3, 7, 9
        | 1, 0 :: _ -> familySize <= 4
        | 3, 0 :: _ -> familySize <= 3
        | 7, 0 :: _ -> familySize <= 2
        | 9, 0 :: _ -> familySize = 1
        | _, 0 :: _ -> failwith "Unexpected branch of match expression"
        // elder position mustn't contain 0
        | digit, positions when positions |> List.contains elderPos -> familySize <= (10 - digit - 1)
        | digit, _ -> familySize <= (10 - digit)

    let getPossibleVariantDigits (elderPos: int) (digitInfo: DigitInfo) =
        match digitInfo.Digit, digitInfo.Positions with
        // last (younger) digit must be 1, 3, 7, 9
        | 1, 0 :: _ -> [1; 3; 7; 9]
        | 3, 0 :: _ -> [3; 7; 9]
        | 7, 0 :: _ -> [7; 9]
        | 9, 0 :: _ -> [9]
        // elder position mustn't contain 0
        | 0, positions when positions |> List.contains elderPos -> [1 .. 9]
        | digit, _ -> [digit .. 9]

    let getPossibleVariants (familySize: int) (digitsInfo : DigitsInfo) =
        digitsInfo.Info |> Seq.filter (fun digitInfo -> digitInfo |> checkPossibleVariant familySize digitsInfo.ElderPos) |> Seq.toList

    let checkPossibleVariant (familySize: int) (sieve: EratosSieve) (digitsInfo: DigitsInfo) (variant : DigitInfo) =
        let numberBasis = digitsInfo |> createNumberBasis variant.Digit
        let rec checkPossibleVariantImpl (suitableVariants: int) (digits: int list) =
            match digits with
            | [] -> false
            | digit :: digitsRest ->
                let number = {variant with DigitInfo.Digit = digit} |> createNumber numberBasis
                match sieve.IsPrime(number) with
                | false -> digitsRest |> checkPossibleVariantImpl suitableVariants
                | true when suitableVariants = (familySize - 1) -> true
                | true -> digitsRest |> checkPossibleVariantImpl (suitableVariants + 1)
        variant |> getPossibleVariantDigits digitsInfo.ElderPos |> checkPossibleVariantImpl 0

    let processNumber (familySize: int) (sieve: EratosSieve) (number: int) =
        let digitsInfo = number |> createDigitsInfo
        digitsInfo |> getPossibleVariants familySize |> Seq.exists (fun variant -> variant |> checkPossibleVariant familySize sieve digitsInfo)

    let solveImpl (familySize: int) =
        let sieve = EratosSieve.Create(MaxNumber)
        // we don't process single digit primes
        sieve.ToSeq() |> Seq.skipWhile (fun prime -> prime < 10) |> Seq.find (fun prime -> prime |> processNumber familySize sieve)

    [<TestCase(6, 13, TimeThresholds.HardTimeLimit)>]
    [<TestCase(7, 56003, TimeThresholds.HardTimeLimit)>]
    [<TestCase(8, 121313, TimeThresholds.HardTimeLimit)>]
    member public this.Solve(familySize: int, expectedAnswer: int, timeLimit: int) =
        SolutionUtils.CheckSolution(timeLimit, expectedAnswer, solveImpl, familySize)
